### ISSUE TITLE
✨ Move selected lines up and down

### ISSUE THREAD

**Issue by austincondiff** [CodeEditApp/CodeEditSourceEditor#259]
### Description

Users should be able to move lines up or down with <kbd>⇧</kbd> <kbd>⌘</kbd> <kbd>[</kbd> and <kbd>]</kbd> or by going to Editor > Move Line Up and Move Line Down.

### Screenshots

<img width="1038" alt="image" src="https://github.com/user-attachments/assets/1cada73e-0b26-48e5-81c3-82807b0018a9">

**Comment by nis-ship-it** [comment]
@austincondiff I would like to work on it, however would need some guidance on where the files located? QQ: Its in CodeEditSourceEditor project?

**Comment by Bogdan-Belogurov** [comment]
@austincondiff Hello! Do you mind if I take this?

**Comment by 0xWDG** [comment]
> [@austincondiff](https://github.com/austincondiff) Hello! Do you mind if I take this?

I have assigned you :)

**Comment by Bogdan-Belogurov** [comment]
@0xWDG
Hello! 
I've made the first prototype
<details><summary>Details</summary>
<p>

<video src="https://github.com/user-attachments/assets/6ece467c-a01d-412e-a856-c02fd8cc38e0" width="46%" />


</p>
</details> 

And I noticed a few problems:

1. If I select a line from the zero position, then in the `editSelections` method, the selection range includes that line, even if the selection is not actually visible in the UI. Is that the way it's meant to be?
<details><summary>Details</summary>
<p>

<video src="https://github.com/user-attachments/assets/6aaf8f99-0996-47fa-94b7-1c7df8607f11" width="46%" />

</p>
</details> 

2. As I see from Xcode, during line up/down actions, the selected lines understand their scope and automatically add/remove additional indents. Can you show me where to look to implement similar logic? 🙏🏿

**Comment by austincondiff** [comment]
First of all, great work so far! Maybe if the selection ends with a newline character, we exclude it. Would that work?

**Comment by thecoolwinter** [comment]
I think you're right to keep the newline character, but it sounds like there might be something going funky with finding the lines being selected to move.

> And I noticed a few problems:
> 
> 1. If I select a line from the zero position, then in the `editSelections` method, the selection range includes that line, even if the selection is not actually visible in the UI. Is that the way it's meant to be?

What are you using to determine what lines are selected? The `editSelections` method just passes along any text selections active in the TextView, so that makes me think there's maybe something going on with whatever method is being used to find lines in the selected range.

> 2. As I see from Xcode, during line up/down actions, the selected lines understand their scope and automatically add/remove additional indents. Can you please show me where to look to implement similar logic? 🙏🏿

Right now we don't have any architecture for determining 'correct' indent levels. I may be building that out right now while developing the line folding gutter and I know `tree-sitter` has that capability, but we haven't had a need for it just yet.

**Comment by Bogdan-Belogurov** [comment]
@thecoolwinter Hello! 
> What are you using to determine what lines are selected? 

I use `getHighlightedLines` method to determine what lines are selected

```swift
guard let lineIndexes = getHighlightedLines(for: selection.range) else { return }
```

Here I've selected starting at line 27 of the zero position 
<img width="233" alt="Image" src="https://github.com/user-attachments/assets/5a8b2f4c-2e3b-49e4-8e3e-422184aa2174" />

getHighlightedLines  output:
<img width="204" alt="Image" src="https://github.com/user-attachments/assets/9eacd919-8657-4081-a20c-233d5c3c7796" />

**Comment by thecoolwinter** [comment]
Okay looking into that method you're right it's not returning what you'd expect. Technically the selection contains the second line (line 27 in your screenshot), but that method shouldn't be returning both those lines. That's actually causing another bug with the indent lines functionality with the same issue you found. I'll have a patch ready in a moment for that method that should make it function as expected.

**Comment by thecoolwinter** [comment]
@Bogdan-Belogurov if you pull from main that method should work as expected now.

**Comment by austincondiff** [comment]
Looks fantastic! One detail. Should it modify your selection? Or should it move the lines, leaving your selection alone? We should follow Xcodes lead on this one.

**Comment by Bogdan-Belogurov** [comment]
@thecoolwinter Hi! Do you need an Editor->Structure menu like in Xcode?

**Comment by thecoolwinter** [comment]
@Bogdan-Belogurov you're right, that's not expected behavior. The coordinator should be setup again after switching back to the old tab. I'll investigate today

**Comment by thecoolwinter** [comment]
@Bogdan-Belogurov got a [fix](https://github.com/CodeEditApp/CodeEditSourceEditor/pull/340) in for the coordinator lifecycle bug, I'll lyk when it's in a tagged release so we can update the related CodeEdit PR.
